contracts/signature_mint/SignatureMint.sol

```diff
@@ -0,0 +1,276 @@
+
+contract SignatureMint is
+    
+    ISignatureMint,
+    ERC721Enumerable,
+    EIP712,
+    AccessControlEnumerable,
+    ERC2771Context,
+    IERC2981,
+    ReentrancyGuard,
+    Multicall
+
+{
+    using ECDSA for bytes32;
+    using Strings for uint256;
+
+    bytes32 private constant TYPEHASH =
+        keccak256("MintRequest(address to,string baseURI,uint256 amountToMint,uint256 validityStartTimestamp,uint256 validityEndTimestamp,bytes uid)");
+
+    /// @dev Only TRANSFER_ROLE holders can have tokens transferred from or to them, during restricted transfers.
+    bytes32 public constant TRANSFER_ROLE = keccak256("TRANSFER_ROLE");
+    /// @dev Only MINTER_ROLE holders can sign off on `MintRequest`s.
+    bytes32 public constant MINTER_ROLE = keccak256("MINTER_ROLE");
+
+    /// @dev The token ID of the next token to mint.
+    uint256 public nextTokenIdToMint;
+
+    /// @dev Contract interprets 10_000 as 100%.
+    uint64 private constant MAX_BPS = 10_000;
+
+    /// @dev The % of secondary sales collected as royalties. See EIP 2981.
+    uint64 public royaltyBps;
+
+    /// @dev The % of primary sales collected by the contract as fees.
+    uint120 public feeBps;
+
+    /// @dev Whether transfers on tokens are restricted.
+    bool public transfersRestricted;
+
+    /// @dev Contract level metadata.
+    string public contractURI;
+    
+    /// @dev The protocol control center.
+    ProtocolControl internal controlCenter;
+
+    uint256[] private baseURIIndices;
+    
+    /// @dev Mapping from end-tokenId => baseURI.
+    mapping(uint256 => string) public baseURI;
+
+    /// @dev Mapping from mint request UID => whether the mint request is processed.
+    mapping(bytes => bool) private minted;
+
+    modifier onlyModuleAdmin() {
+        require(hasRole(DEFAULT_ADMIN_ROLE, _msgSender()), "not module admin.");
+        _;
+    }
+    
+    constructor(
+        string memory _name,
+        string memory _symbol,
+        string memory _contractURI,
+        address payable _controlCenter,
+        address _trustedForwarder,
+        uint128 _royaltyBps,
+        uint128 _feeBps
+    ) 
+        ERC721(_name, _symbol) 
+        EIP712(_name, "1")
+        ERC2771Context(_trustedForwarder)
+    {
+        // Set the protocol control center
+        controlCenter = ProtocolControl(_controlCenter);
+        contractURI = _contractURI;
+        royaltyBps = uint64(_royaltyBps);
+        feeBps = uint120(_feeBps);
+
+        address deployer = _msgSender();
+        _setupRole(DEFAULT_ADMIN_ROLE, deployer);
+        _setupRole(MINTER_ROLE, deployer);
+    }
+
+    ///     =====   Public functions  =====
+
+    /// @dev Verifies that a mint request is signed by an account holding MINTER_ROLE (at the time of the function call).
+    function verify(
+        MintRequest calldata req,
+        bytes calldata signature
+    )
+        public
+        view 
+        returns (bool)
+    {
+        address signer = _hashTypedDataV4(
```

Comment:
is `chainid` taken care of in the recover signature? 
basically don't allow chainid replay attack

Reply:
Good point! I think we should have the contract make this check anyway.